File: web/src/tiles/infinitePowerSource.ts
import {Tile} from './base';

class InfinitePowerSourceTile extends Tile {
    TILE_TYPE = 'INFINITE POWER SOURCE';
    SAMPLE_CHARS = ["⏻"]
    POWER = 255;

    constructor(ctx: ExecutionContext, x: number, y: number) {
        super(ctx, x, y);
    }

    startTurn() {
        for (let dx = 0; dx < 2; dx++) {
            for (let dy = 0; dy < 2; dy++) {
                if ((dx == 0 && dy == 0) || (dx != 0 && dy != 0)) {
                    continue;
                }
                const targetX = this.x + dx;
                const targetY = this.y + dy;
                const targetTile = this.ctx.getTile(targetX, targetY);
                if (targetTile) {
                    targetTile.acceptEnergy(this.POWER, this.x, this.y);
                }
            }
        }
    }
}

File: lectrick/tiles/infinite_power_source.py
from .base import Tile


class InfinitePowerSource(Tile):
    TILE_TYPE = "INFINITE POWER SOURCE"
    SAMPLE_CHARS = ["⏻"]
    POWER_LEVEL = 255

    def __init__(self, x, y, tiles):
        super().__init__(x, y, tiles)

    def start_turn(self, *args, **kwargs):
        for dx in range(-1, 2):
            for dy in range(-1, 2):
                if (dx == 0 and dy == 0) or (dx != 0 and dy != 0):
                    continue
                target_x = self.x + dx
                target_y = self.y + dy
                try:
                    tile_index = self.tiles.index((target_x, target_y))
                except ValueError:
                    continue
                self.tiles[tile_index].accept_energy(self.POWER_LEVEL, self.x, self.y)
